### PR TITLE
Adding devcontainer for easy Python development

### PR DESCRIPTION
**Pull request by storey247**
Love the tool, wanted to contribute and here is the result :)

- Added a [VSCode devcontainer](https://code.visualstudio.com/docs/remote/containers) to help people get up and running with the repo quicker without needing to install any tooling or do any setup of venv
- Added a new rule to help when some repositories use `main` and some use `master` as the primary integration branch

Feedback welcome and thanks for the great tooling

### CODE REVIEW

File: thefuck/rules/git_main_master.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from thefuck.specific.git import git_support
+
+
+@git_support
+def match(command):
+    return "'master'" in command.output.lower() or "'main'" in command.output.lower()
```
Comment (scorphus) [inline thefuck/rules/git_main_master.py:6]: Why is `.lower()` needed here?

File: thefuck/rules/git_main_master.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+
+@git_support
+def get_new_command(command):
+    if "'master'" in command.output.lower():
+        return command.script.replace("master", "main")
+    else:
+        return command.script.replace("main", "master")
```
Comment (scorphus) [inline thefuck/rules/git_main_master.py:14]:
```suggestion
    return command.script.replace("main", "master")
```

**Review by scorphus — CHANGES_REQUESTED** [review]
Looking good!

Would you be willing to separate the two features into two different commits and — if it's not too much to ask — two different PR's?

Thanks!

**Comment by scorphus** [comment]
Also, the fix for Python 2.7 tests is already on current `master`, would you please rebase on top of that? 🙂

**Comment by scorphus** [comment]
Howdy @storey247! Thanks for contributing. I'd really love to read your comments on my suggestions above. I understand that you're probably busy with many stuff. But I thought I'd ping you 🙂

**Comment by storey247** [comment]
hi @scorphus thanks for all the feedback, yes I will action and get this PR split into two :smile: apologies for the delayed response, work is a bit crazy atm

File: .devcontainer/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# See here for image contents: https://github.com/microsoft/vscode-dev-containers/tree/v0.163.1/containers/python-3/.devcontainer/base.Dockerfile
+
+# [Choice] Python version: 3, 3.9, 3.8, 3.7, 3.6
+ARG VARIANT="3"
+FROM mcr.microsoft.com/vscode/devcontainers/python:0-${VARIANT}
+
+# [Option] Install Node.js
+ARG INSTALL_NODE="true"
+ARG NODE_VERSION="lts/*"
+RUN if [ "${INSTALL_NODE}" = "true" ]; then su vscode -c "umask 0002 && . /usr/local/share/nvm/nvm.sh && nvm install ${NODE_VERSION} 2>&1"; fi
```
[the remark this answers is not in the document]
Comment (storey247) [reply .devcontainer/Dockerfile:10]: the devconainer.json file actually set the arg to false, so it's not getting installed, but as it's not necessary to have this I will remove 👍

File: .devcontainer/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+# COPY requirements.txt /tmp/pip-tmp/
+# RUN pip3 --disable-pip-version-check --no-cache-dir install -r /tmp/pip-tmp/requirements.txt \
+#    && rm -rf /tmp/pip-tmp
+
+# [Optional] Uncomment this section to install additional OS packages.
+# RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
+#     && apt-get -y install --no-install-recommends <your-package-list-here>
+
+# [Optional] Uncomment this line to install global node packages.
+# RUN su vscode -c "source /usr/local/share/nvm/nvm.sh && npm install -g <your-package-here>" 2>&1
```
[the remark this answers is not in the document]
Comment (storey247) [reply .devcontainer/Dockerfile]: will clean up

File: .devcontainer/devcontainer.json [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+	// Use 'forwardPorts' to make a list of ports inside the container available locally.
+	// "forwardPorts": [],
+
+	// Use 'postCreateCommand' to run commands after the container is created.
+	"postCreateCommand": "pip3 install -r requirements.txt && python3 setup.py develop",
+
+	// Comment out connect as root instead. More info: https://aka.ms/vscode-remote/containers/non-root.
+	//"remoteUser": "vscode"
+}
```
[the remark this answers is not in the document]
Comment (storey247) [reply .devcontainer/devcontainer.json:47]: again, will clean

**Comment by scorphus** [comment]
Awesome! I already split the commit into three, so no need to create another PR, I can merge the changes regarding the new rule and we leave this PR for the devcontainer addition. What do you think?

**Comment by storey247** [comment]
Amazing! Nice work! 👏👏👏

sounds like a plan, I’ll rebase this on Monday when the work is merged and I’ll tidy up the container so it’s nice and trim. 

Thanks @scorphus

**Comment by scorphus** [comment]
There you go, @storey247. Sorry for the extra noise. To continue your work, I'd suggest you force-resetting to the `main_master` branch on your remote, that way you'll have exactly what this PR currently has. Please don't hesitate to ask me for help with that.

**Comment by storey247** [comment]
@scorphus I have now rebased my work from latest `master` and cleaned up the comments as discussed.

I also added some notes into the `Contribute.md` to help people get up and running with the devcontainer setup if required.

Hopefully now this work can be merged in too. Let me know if there is anything else you need me to look at

File: .devcontainer/Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (storey247) [reply .devcontainer/Dockerfile]: I have also cleaned all this up as node is not required

**Review by scorphus — APPROVED** [review]
Thats super, @storey247! Thanks so much for hanging in! Please consider my suggestion below.

File: .devcontainer/devcontainer.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
+// https://github.com/microsoft/vscode-dev-containers/tree/v0.163.1/containers/python-3
+{
+	"name": "Python 3",
+	"build": {
+		"dockerfile": "Dockerfile",
+		"context": "..",
+		"args": { 
+			// Update 'VARIANT' to pick a Python version: 3, 3.6, 3.7, 3.8, 3.9
+			"VARIANT": "3.9"
```
[the remark this answers is not in the document]
Comment (storey247) [reply .devcontainer/devcontainer.json:10]: yes you are correct, the ARG will be overridden by the value passed in the `devcontainer.json` and acts like a fallback default. Will set the `VARIANT` to be 3 as suggested

[the remark this answers is not in the document]
Comment (storey247) [reply .devcontainer/devcontainer.json]: done and sync'd to remote

**Comment by storey247** [comment]
> Thats super, @storey247! Thanks so much for hanging in! Please consider my suggestion below.

Feedback actions, just waiting on the builds and then will merge 👍

File: .devcontainer/devcontainer.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+// https://github.com/microsoft/vscode-dev-containers/tree/v0.163.1/containers/python-3
+{
+	"name": "Python 3",
+	"build": {
+		"dockerfile": "Dockerfile",
+		"context": "..",
+		"args": { 
+			// Update 'VARIANT' to pick a Python version: 3, 3.6, 3.7, 3.8, 3.9
+			"VARIANT": "3"
+		}
```
Comment (scorphus) [inline .devcontainer/devcontainer.json:11]:
```suggestion
		"context": ".."
```
Let's remove that redundancy 🙂

File: .devcontainer/devcontainer.json [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+	},
+
+	// Set *default* container specific settings.json values on container create.
+	"settings": {
```
Comment (scorphus) [inline .devcontainer/devcontainer.json:15]:
```suggestion
	"settings": {
```

File: .devcontainer/devcontainer.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+		"python.linting.pylintPath": "/usr/local/py-utils/bin/pylint"
+	},
+
+	// Add the IDs of extensions you want installed when the container is created.
+	"extensions": [
+		"ms-python.python"
+	],
+
+	// Use 'postCreateCommand' to run commands after the container is created.
+	"postCreateCommand": "pip3 install -r requirements.txt && python3 setup.py develop",
```
Comment (scorphus) [inline .devcontainer/devcontainer.json:37]:
```suggestion
	"postCreateCommand": "pip3 install -r requirements.txt && python3 setup.py develop"
```

**Review by scorphus — APPROVED** [review]
Thanks, @storey247! That's a great contribution!

Also, thanks for the new `git_main_master` rule! 🙌

**Pull request merged**